gapis/trace/desktop/trace.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package desktop
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"runtime"
+
+	"github.com/google/gapid/core/os/device/bind"
+	"github.com/google/gapid/core/os/process"
+	"github.com/google/gapid/core/vulkan/loader"
+	gapii "github.com/google/gapid/gapii/client"
+	"github.com/google/gapid/gapis/trace/tracer"
+)
+
+// Only update the package list every 30 seconds at most
+var packageUpdateTime = 30.0
+
+type DesktopTracer struct {
+	b bind.Device
+}
+
+func NewTracer(dev bind.Device) *DesktopTracer {
+	return &DesktopTracer{dev.(bind.Device)}
+}
+
+// IsServerLocal returns true if all paths on this device can be server-local
+func (t *DesktopTracer) IsServerLocal() bool {
+	return true
+}
+
+func (t *DesktopTracer) CanSpecifyCWD() bool {
+	return true
+}
+
+func (t *DesktopTracer) CanSpecifyEnv() bool {
+	return true
+}
+
+func (t *DesktopTracer) CanUploadApplication() bool {
+	return false
+}
+
+func (t *DesktopTracer) HasCache() bool {
+	return false
+}
+
+func (t *DesktopTracer) CanUsePortFile() bool {
+	return true
+}
+
+func (t *DesktopTracer) APITraceOptions() []tracer.APITraceOptions {
+	options := make([]tracer.APITraceOptions, 0, 1)
+	if len(t.b.Instance().Configuration.Drivers.Vulkan.PhysicalDevices) > 0 {
+		options = append(options, tracer.VulkanTraceOptions())
+	}
+	return options
+}
+
+func (t *DesktopTracer) FindTraceTarget(ctx context.Context, str string) (*tracer.TraceTargetTreeNode, error) {
+	isFile, err := t.b.IsFile(ctx, str)
+	if err != nil {
+		return nil, err
+	}
+	if !isFile {
+		return nil, fmt.Errorf("Trace target is not an executable file %+v", str)
+	}
+	dir, file := filepath.Split(str)
+
+	if dir == "" {
+		dir = "."
+		if runtime.GOOS == "windows" {
+			str = ".\\" + file
+		} else {
+			str = "./" + file
+		}
+	}
+
+	tttn := &tracer.TraceTargetTreeNode{
+		file,
```

Comment:
Use named fields when constructing a struct declared in another package.

Reply:
Done